hw_5_pandas_code/hw_5_pandas.py

```diff
@@ -0,0 +1,239 @@
+rna_count_table = rna_count_table.reset_index(level=0).rename({'index' : 'Sequence'}, axis = 'columns')
+
+#rna_final_table = pd.melt(rna_count_table, id_vars='Sequence', var_name='rRNA', value_name='Count')
+#rna_final_table.head()
+
+plt.style.use('default')
+rna_count_table.plot(x='Sequence', y=['16S', '23S', '5S'], kind='bar', figsize=(20, 10))
+plt.ylabel('Count')
+plt.legend(title='RNA type')
+
+rna_annotation['interval_ann'] = rna_annotation.apply(lambda row: (row.start, row.end), axis = 1)
+alignment['interval_align'] = alignment.apply(lambda row: (row.start - 1, row.end + 1), axis = 1)
+
+# Function to check if any two intervals overlap
+def intersect(first: tuple, second: tuple) -> bool:
+    if first[0] >= second[0] and first[1] <= second[1]:
+            return True
+    return False
+
+df_merge = rna_annotation.merge(alignment, on='chromosome')
+df_merge = df_merge[df_merge.apply(lambda row : intersect(row.interval_ann, row.interval_align), axis = 1)]
+df_merge.drop(['interval_ann', 'interval_align'], axis=1, inplace=True)
+
+df_merge
+
+"""## Задание 2. Кастомизация графиков"""
+
+from matplotlib.ticker import AutoMinorLocator
+import heapq
+
+volcano_data = pd.read_csv('data/diffexpr_data.tsv.gz', sep = '\t')
+volcano_data
+
+def vol_col(table: pd.DataFrame) -> str:
+    logfc, pval = table
+    if logfc < 0 and pval < 0.05:
+        return 'Significantly down-regulated'
+    if logfc < 0 and pval > 0.05:
+        return 'Non-significantly down-regulated'
+    if logfc > 0 and pval < 0.05:
+        return 'Significantly up-regulated'
+    return 'Non-significantly up-regulated'
+
+volcano_data['regulation'] = volcano_data[['logFC', 'pval_corr']].apply(vol_col, axis=1)
+
+# Setting palette for the legend
+palette = {'Significantly down-regulated': '#377eb8', 'Significantly up-regulated': '#ff7f00', 'Non-significantly down-regulated': '#4daf4a', 'Non-significantly up-regulated': '#e41a1c'}
+
+# Setting x-limits
+min_max = abs(min(volcano_data['logFC']) - 1), abs(max(volcano_data['logFC']) + 1)
+max_val = max(min_max)
+sim_bord = [-max_val, max_val]
+
+# Adding annotation to top genes
+def get_samples(dataframe: pd.DataFrame):
+    up_dataframe = dataframe.loc[(dataframe['regulation'] == 'Significantly up-regulated') & (dataframe['pval_corr'] < 0.05)]
+    down_dataframe = dataframe.loc[(dataframe['regulation'] == 'Significantly down-regulated') & (dataframe['pval_corr'] < 0.05)]
+    reg = heapq.nlargest(2, up_dataframe.logFC) + [-x for x in heapq.nlargest(2, abs(down_dataframe.logFC))]
+    samples = []
+    pval = []
+    for val in reg:
+        float_val = float(val)
+        samples.append(dataframe.loc[dataframe['logFC'] == float_val, 'Sample'].iloc[0])
+        pval.append(dataframe.loc[dataframe['logFC'] == float_val, 'log_pval'].iloc[0])
+    return reg, samples, pval
+
+reg, samples_exp, log_pval = get_samples(volcano_data)
```

Comment:
```suggestion
top_upregulated = volcano_data.query("pval_corr < 0.05").nlargest(2, "logFC")
top_downregulated = volcano_data.query("pval_corr < 0.05").nsmallest(2, "logFC")

```
![image](https://user-images.githubusercontent.com/48160953/206874598-cc33cf38-fc1b-4908-93f4-89e3bb3c44d5.png)